Error creating new record

It happens in test environment with survey "SHARP+ 2020 (DEMO VERSION)".

When a new record is created, there is an error server side: 
```
[2020-10-30T00:41:26.281] [ERROR] arena - Thread - Error in thread:  BatchError {
    stat: { total: 101, succeeded: 18, failed: 83, duration: 92 }
    errors: [
        100: error: current transaction is aborted, commands ignored until end of transaction block
            at Parser.parseErrorMessage (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/parser.js:278:15)
            at Parser.handlePacket (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/parser.js:126:29)
            at Parser.parse (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/parser.js:39:38)
            at Socket.<anonymous> (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/index.js:8:42)
            at Socket.emit (events.js:315:20)
            at addChunk (_stream_readable.js:295:12)
            at readableAddChunk (_stream_readable.js:271:9)
            at Socket.Readable.push (_stream_readable.js:212:10)
            at TCP.onStreamRead (internal/stream_base_commons.js:186:23)
        From previous event:
            at promise (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/promise-parser.js:30:20)
            at Task.$query (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/query.js:145:12)
            at Task.<anonymous> (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/query.js:263:23)
            at Task.query (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/task.js:117:34)
            at Object._update [as update] (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:16466:44)
            at /home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:16485:68
            at Array.map (<anonymous>)
            at Module.updateTable (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:16485:30)
            at _onNodesUpdate (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:14941:118)
            at async Task.<anonymous> (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:14890:7) {
          length: 144,
          severity: 'ERROR',
          code: '25P02',
          detail: undefined,
          hint: undefined,
          position: undefined,
          internalPosition: undefined,
          internalQuery: undefined,
          where: undefined,
          schema: undefined,
          table: undefined,
          column: undefined,
          dataType: undefined,
          constraint: undefined,
          file: 'postgres.c',
          line: '1105',
          routine: 'exec_simple_query'
        }
    ]
}
[2020-10-30T00:41:26.281] [ERROR] arena - Thread - BatchError {
    stat: { total: 101, succeeded: 18, failed: 83, duration: 92 }
    errors: [
        100: error: current transaction is aborted, commands ignored until end of transaction block
            at Parser.parseErrorMessage (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/parser.js:278:15)
            at Parser.handlePacket (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/parser.js:126:29)
            at Parser.parse (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/parser.js:39:38)
            at Socket.<anonymous> (/home/ste/dev/projects/openforis/arena/node_modules/pg-protocol/dist/index.js:8:42)
            at Socket.emit (events.js:315:20)
            at addChunk (_stream_readable.js:295:12)
            at readableAddChunk (_stream_readable.js:271:9)
            at Socket.Readable.push (_stream_readable.js:212:10)
            at TCP.onStreamRead (internal/stream_base_commons.js:186:23)
        From previous event:
            at promise (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/promise-parser.js:30:20)
            at Task.$query (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/query.js:145:12)
            at Task.<anonymous> (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/query.js:263:23)
            at Task.query (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/task.js:117:34)
            at Object._update [as update] (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:16466:44)
            at /home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:16485:68
            at Array.map (<anonymous>)
            at Module.updateTable (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:16485:30)
            at _onNodesUpdate (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:14941:118)
            at async Task.<anonymous> (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:14890:7) {
          length: 144,
          severity: 'ERROR',
          code: '25P02',
          detail: undefined,
          hint: undefined,
          position: undefined,
          internalPosition: undefined,
          internalQuery: undefined,
          where: undefined,
          schema: undefined,
          table: undefined,
          column: undefined,
          dataType: undefined,
          constraint: undefined,
          file: 'postgres.c',
          line: '1105',
          routine: 'exec_simple_query'
        }
    ]
}
    at check (/home/ste/dev/projects/openforis/arena/node_modules/spex/lib/ext/batch.js:135:32)
    at step (/home/ste/dev/projects/openforis/arena/node_modules/spex/lib/ext/batch.js:109:17)
    at /home/ste/dev/projects/openforis/arena/node_modules/spex/lib/ext/batch.js:83:17
    at /home/ste/dev/projects/openforis/arena/node_modules/spex/lib/utils/index.js:62:25
    at AsyncResource.runInAsyncScope (async_hooks.js:186:9)
    at processImmediate (internal/timers.js:456:21)
From previous event:
    at promise (/home/ste/dev/projects/openforis/arena/node_modules/spex/lib/index.js:91:24)
    at Task.batch (/home/ste/dev/projects/openforis/arena/node_modules/spex/lib/ext/batch.js:73:12)
    at Task.<anonymous> (/home/ste/dev/projects/openforis/arena/node_modules/spex/lib/ext/batch.js:149:22)
    at Task.batch (/home/ste/dev/projects/openforis/arena/node_modules/pg-promise/lib/task.js:135:39)
    at Module.updateTable (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:16485:16)
    at _onNodesUpdate (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:14941:118)
    at async Task.<anonymous> (/home/ste/dev/projects/openforis/arena/dist/recordUpdateThread.js:14890:7)

```

--- a/server/modules/surveyRdb/repository/dataTableUpdateRepository.js
+++ b/server/modules/surveyRdb/repository/dataTableUpdateRepository.js
@@ -1,5 +1,3 @@
-import * as R from 'ramda'
-
 import * as Survey from '@core/survey/survey'
 import * as NodeDef from '@core/survey/nodeDef'
 import * as Node from '@core/record/node'
@@ -19,15 +17,19 @@ const _getType = (nodeDef, node) => {
   const updated = Node.isUpdated(node)
   const deleted = Node.isDeleted(node)
 
-  return created && _hasTable(nodeDef)
-    ? types.insert
-    : updated || created
-    ? types.update
-    : deleted && _hasTable(nodeDef)
-    ? types.delete
-    : deleted
-    ? types.update
-    : null
+  if (created && _hasTable(nodeDef)) {
+    return types.insert
+  }
+  if (updated || created) {
+    return types.update
+  }
+  if (deleted && _hasTable(nodeDef)) {
+    return types.delete
+  }
+  if (deleted) {
+    return types.update
+  }
+  return null
 }
 
 const _getColNames = (nodeDef, type) =>
@@ -43,7 +45,7 @@ const _getColNames = (nodeDef, type) =>
       ]
     : DataCol.getNames(nodeDef)
 
-const _getColValues = async (survey, cycle, nodeDef, node, type) =>
+const _getColValues = (survey, cycle, nodeDef, node, type) =>
   type === types.insert
     ? [
         Node.getUuid(node),
@@ -51,33 +53,34 @@ const _getColValues = async (survey, cycle, nodeDef, node, type) =>
         cycle,
         Node.getParentUuid(node),
         ...(NodeDef.isMultipleAttribute(nodeDef) // Entity
-          ? await Promise.all(DataCol.getValues(survey, nodeDef, node))
+          ? DataCol.getValues(survey, nodeDef, node)
           : []),
       ]
-    : await DataCol.getValues(survey, nodeDef, node)
+    : DataCol.getValues(survey, nodeDef, node)
 
 const _getRowUuid = (nodeDef, node, nodeParent) => (_hasTable(nodeDef) ? Node.getUuid(node) : Node.getUuid(nodeParent))
 
-const _toUpdates = async (survey, cycle, nodeDefs, nodes) => {
-  const updates = await Promise.all(
-    R.values(nodes).map(async node => {
-      const nodeDef = nodeDefs[Node.getNodeDefUuid(node)]
-      const nodeDefParent = nodeDefs[NodeDef.getParentUuid(nodeDef)]
-      const type = _getType(nodeDef, node)
-
-      return type
-        ? {
-            type,
-            schemaName: SchemaRdb.getName(Survey.getId(survey)),
-            tableName: DataTable.getName(nodeDef, nodeDefParent),
-            colNames: _getColNames(nodeDef, type),
-            colValues: await _getColValues(survey, cycle, nodeDef, node, type),
-            rowUuid: _getRowUuid(nodeDef, node, nodes[Node.getParentUuid(node)]),
-          }
-        : null
-    }),
+const _toUpdates = (survey, cycle, nodeDefs, nodes) => {
+  // visit nodes with BFS algorithm to avoid FK constraints violations (sort nodes by hierarchy depth)
+  const nodesArray = Object.values(nodes).sort(
+    (nodeA, nodeB) => Node.getHierarchy(nodeA).length - Node.getHierarchy(nodeB).length
   )
-  return R.reject(R.isNil, updates)
+  return nodesArray.reduce((updatesAcc, node) => {
+    const nodeDef = nodeDefs[Node.getNodeDefUuid(node)]
+    const nodeDefParent = nodeDefs[NodeDef.getParentUuid(nodeDef)]
+    const type = _getType(nodeDef, node)
+    if (type) {
+      updatesAcc.push({
+        type,
+        schemaName: SchemaRdb.getName(Survey.getId(survey)),
+        tableName: DataTable.getName(nodeDef, nodeDefParent),
+        colNames: _getColNames(nodeDef, type),
+        colValues: _getColValues(survey, cycle, nodeDef, node, type),
+        rowUuid: _getRowUuid(nodeDef, node, nodes[Node.getParentUuid(node)]),
+      })
+    }
+    return updatesAcc
+  }, [])
 }
 
 // ==== execution
@@ -87,7 +90,7 @@ const _update = (update, client) =>
     `UPDATE ${update.schemaName}.${update.tableName}
       SET ${update.colNames.map((col, i) => `${col} = $${i + 2}`).join(',')}
       WHERE uuid = $1`,
-    [update.rowUuid, ...update.colValues],
+    [update.rowUuid, ...update.colValues]
   )
 
 const _insert = (update, client) =>
@@ -96,7 +99,7 @@ const _insert = (update, client) =>
       (${update.colNames.join(',')})
       VALUES 
       (${update.colNames.map((_col, i) => `$${i + 1}`).join(',')})`,
-    update.colValues,
+    update.colValues
   )
 
 const _delete = (update, client) =>
@@ -109,7 +112,6 @@ const queryByType = {
 }
 
 export const updateTable = async (survey, cycle, nodeDefs, nodes, client) => {
-  const updates = await _toUpdates(survey, cycle, nodeDefs, nodes)
-
-  await client.batch(updates.map(update => queryByType[update.type](update, client)))
+  const updates = _toUpdates(survey, cycle, nodeDefs, nodes)
+  await client.batch(updates.map((update) => queryByType[update.type](update, client)))
 }

--- a/server/modules/surveyRdb/repository/dataView/read.js
+++ b/server/modules/surveyRdb/repository/dataView/read.js
@@ -262,7 +262,7 @@ export const fetchRecordsCountByKeys = async (
   const nodeDefKeys = Survey.getNodeDefKeys(nodeDefRoot)(survey)
   const surveyId = Survey.getId(survey)
   const schemaRdb = SchemaRdb.getName(surveyId)
-  const schema = Schemata.getSchemaSurvey(survey)
+  const schema = Schemata.getSchemaSurvey(surveyId)
   const rootTable = `${schemaRdb}.${NodeDefTable.getViewName(nodeDefRoot)}`
   const rootTableAlias = 'r'
 

